File: backend/encryption.py
import os
import base64
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives import padding
from cryptography.hazmat.backends import default_backend

# Key size for AES encryption (256 bits)
KEY_SIZE = 32
# IV size for AES encryption (128 bits)
IV_SIZE = 16

def generate_encryption_key():
    """Generate a random encryption key"""
    return os.urandom(KEY_SIZE)

def generate_iv():
    """Generate a random initialization vector (IV)"""
    return os.urandom(IV_SIZE)

def encrypt_message(message, key, iv=None):
    """
    Encrypt a message using AES-256-CBC
    
    Args:
        message (str): The message to encrypt
        key (bytes): The encryption key
        iv (bytes, optional): The initialization vector. If None, a random IV will be generated.
        
    Returns:
        dict: A dictionary containing the encrypted data and IV
    """
    try:
        # Convert message to bytes if it's a string
        if isinstance(message, str):
            message = message.encode('utf-8')
            
        # Generate a random IV if none is provided
        if iv is None:
            iv = generate_iv()
            
        # Pad the message to a multiple of the block size
        padder = padding.PKCS7(algorithms.AES.block_size).padder()
        padded_message = padder.update(message) + padder.finalize()
        
        # Create an encryptor
        cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=default_backend())
        encryptor = cipher.encryptor()
        
        # Encrypt the message
        encrypted_data = encryptor.update(padded_message) + encryptor.finalize()
        
        # Return the encrypted data and IV as base64 strings
        return {
            'encrypted_data': base64.b64encode(encrypted_data).decode('utf-8'),
            'iv': base64.b64encode(iv).decode('utf-8')
        }
    except Exception as e:
        print(f"Encryption error: {e}")
        return None

def decrypt_message(encrypted_data, iv, key):
    """
    Decrypt a message using AES-256-CBC
    
    Args:
        encrypted_data (str): Base64 encoded encrypted data
        iv (str): Base64 encoded initialization vector
        key (bytes): The encryption key
        
    Returns:
        str: The decrypted message
    """
    try:
        # Convert base64 strings to bytes
        encrypted_data = base64.b64decode(encrypted_data)
        iv = base64.b64decode(iv)
        
        # Create a decryptor
        cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=default_backend())
        decryptor = cipher.decryptor()
        
        # Decrypt the message
        padded_message = decryptor.update(encrypted_data) + decryptor.finalize()
        
        # Unpad the message
        unpadder = padding.PKCS7(algorithms.AES.block_size).unpadder()
        message = unpadder.update(padded_message) + unpadder.finalize()
        
        # Return the decrypted message as a string
        return message.decode('utf-8')
    except Exception as e:
        print(f"Decryption error: {e}")
        return None

def generate_key_pair():
    """
    Generate a key pair for key exchange
    This is a simplified implementation for demonstration purposes
    In a real application, you would use a proper asymmetric encryption algorithm
    
    Returns:
        tuple: A tuple containing the public and private keys
    """
    private_key = os.urandom(KEY_SIZE)
    # In a real implementation, this would be derived from the private key
    # using a proper asymmetric algorithm
    public_key = os.urandom(KEY_SIZE)
    return public_key, private_key

def generate_shared_secret(public_key, private_key):
    """
    Generate a shared secret key using a key exchange protocol
    This is a simplified implementation for demonstration purposes
    In a real application, you would use a proper key exchange protocol
    
    Args:
        public_key (bytes): The other party's public key
        private_key (bytes): Your private key
        
    Returns:
        bytes: The shared secret key
    """
    # In a real implementation, this would be a proper key exchange calculation
    # For now, we'll just combine the keys and hash them
    import hashlib
    combined = public_key + private_key
    hash_obj = hashlib.sha256(combined)
    return hash_obj.digest()

File: src/utils/encryption.js
import CryptoJS from 'crypto-js';

// Key size for AES encryption (256 bits)
const KEY_SIZE = 32;
// IV size for AES encryption (128 bits)
const IV_SIZE = 16;

/**
 * Generate a random encryption key
 * @returns {string} Base64 encoded encryption key
 */
export const generateEncryptionKey = () => {
  const randomBytes = new Uint8Array(KEY_SIZE);
  window.crypto.getRandomValues(randomBytes);
  return arrayBufferToBase64(randomBytes);
};

/**
 * Generate a random initialization vector (IV)
 * @returns {string} Base64 encoded IV
 */
export const generateIV = () => {
  const randomBytes = new Uint8Array(IV_SIZE);
  window.crypto.getRandomValues(randomBytes);
  return arrayBufferToBase64(randomBytes);
};

/**
 * Convert ArrayBuffer to Base64 string
 * @param {ArrayBuffer} buffer - The array buffer to convert
 * @returns {string} Base64 encoded string
 */
export const arrayBufferToBase64 = (buffer) => {
  const bytes = new Uint8Array(buffer);
  let binary = '';
  for (let i = 0; i < bytes.byteLength; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return window.btoa(binary);
};

/**
 * Convert Base64 string to ArrayBuffer
 * @param {string} base64 - The Base64 string to convert
 * @returns {ArrayBuffer} The resulting array buffer
 */
export const base64ToArrayBuffer = (base64) => {
  const binaryString = window.atob(base64);
  const bytes = new Uint8Array(binaryString.length);
  for (let i = 0; i < binaryString.length; i++) {
    bytes[i] = binaryString.charCodeAt(i);
  }
  return bytes.buffer;
};

/**
 * Encrypt a message using AES-256-GCM
 * @param {string} message - The message to encrypt
 * @param {string} key - Base64 encoded encryption key
 * @returns {Object} Object containing encrypted data and IV
 */
export const encryptMessage = (message, key) => {
  try {
    // Generate a random IV for this message
    const iv = generateIV();
    
    // Convert key and IV from Base64 to WordArray
    const keyWordArray = CryptoJS.enc.Base64.parse(key);
    const ivWordArray = CryptoJS.enc.Base64.parse(iv);
    
    // Encrypt the message
    const encrypted = CryptoJS.AES.encrypt(message, keyWordArray, {
      iv: ivWordArray,
      mode: CryptoJS.mode.CBC,
      padding: CryptoJS.pad.Pkcs7
    });
    
    // Return the encrypted message and IV
    return {
      encryptedData: encrypted.toString(),
      iv: iv
    };
  } catch (error) {
    console.error('Encryption error:', error);
    return null;
  }
};

/**
 * Decrypt a message using AES-256-GCM
 * @param {string} encryptedData - The encrypted message
 * @param {string} iv - Base64 encoded IV
 * @param {string} key - Base64 encoded encryption key
 * @returns {string} The decrypted message
 */
export const decryptMessage = (encryptedData, iv, key) => {
  try {
    // Convert key and IV from Base64 to WordArray
    const keyWordArray = CryptoJS.enc.Base64.parse(key);
    const ivWordArray = CryptoJS.enc.Base64.parse(iv);
    
    // Decrypt the message
    const decrypted = CryptoJS.AES.decrypt(encryptedData, keyWordArray, {
      iv: ivWordArray,
      mode: CryptoJS.mode.CBC,
      padding: CryptoJS.pad.Pkcs7
    });
    
    // Convert the decrypted data to a string
    return decrypted.toString(CryptoJS.enc.Utf8);
  } catch (error) {
    console.error('Decryption error:', error);
    return null;
  }
};

/**
 * Generate a shared secret key using Diffie-Hellman key exchange
 * This is a simplified implementation for demonstration purposes
 * In a real application, you would use a proper Diffie-Hellman implementation
 * @param {string} publicKey - The other party's public key
 * @param {string} privateKey - Your private key
 * @returns {string} The shared secret key
 */
export const generateSharedSecret = (publicKey, privateKey) => {
  // In a real implementation, this would be a proper Diffie-Hellman key exchange
  // For now, we'll just combine the keys and hash them
  const combined = publicKey + privateKey;
  const hash = CryptoJS.SHA256(combined);
  return hash.toString(CryptoJS.enc.Base64);
};

/**
 * Store encryption keys securely in localStorage
 * @param {string} contactId - The contact ID
 * @param {string} key - The encryption key
 */
export const storeEncryptionKey = (contactId, key) => {
  const encryptionKeys = JSON.parse(localStorage.getItem('encryptionKeys') || '{}');
  encryptionKeys[contactId] = key;
  localStorage.setItem('encryptionKeys', JSON.stringify(encryptionKeys));
};

/**
 * Retrieve an encryption key from localStorage
 * @param {string} contactId - The contact ID
 * @returns {string|null} The encryption key or null if not found
 */
export const getEncryptionKey = (contactId) => {
  const encryptionKeys = JSON.parse(localStorage.getItem('encryptionKeys') || '{}');
  return encryptionKeys[contactId] || null;
};
